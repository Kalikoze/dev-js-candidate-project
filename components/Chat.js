import React, { Component } from 'react';
import fetch from 'isomorphic-unfetch';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { postMessage } from '../store';
import PropTypes from 'prop-types';

class Chat extends Component {
  constructor() {
    super();
    this.state = {
      input: ''
    };
  }

  async submit(e) {
    if (e.keyCode === 13 || e.button === 0) {
      const { input } = this.state;
      const { addMessage, postMessage } = this.props;

      addMessage(input);
      await postMessage(input);

      const { message, intent } = this.props.response;

      addMessage(message, intent || '#dont_understand')
      this.setState({input: ''});
    }
  }

  render() {
    const { input } = this.state;
    return (
      <section className='input-section'>
          <input
            placeholder='Chat with SpruceBot here...'
            value={input}
            onChange={e => this.setState({input: e.target.value})}
            onKeyDown={e => this.submit(e)}
          />
          <button onClick={e => this.submit(e)}></button>
        <style jsx>{`
          .input-section {
            align-items: center;
            display: flex;
            height: 15%;
            width: 70%;
          }

          input {
            border: none;
            border-radius: 25px;
            font-family: 'Lato', sans-serif;
            font-size: 28px;
            height: 50px;
            padding-left: 25px;
            width: 100%;
          }

          input:focus {
            outline: none;
          }

          button {
            background : url('../static/right-arrow.svg') #4B9DF6 no-repeat center / 50%;
            border: none;
            border-radius: 50%;
            cursor: pointer;
            height: 40px;
            outline: none;
            position: absolute;
            right: 15.7%;
            width: 40px;
          }

          button:active {
            background-color: #8EBBF4;
          }
        `}</style>
      </section>
    );
  }
}

Chat.propTypes = {
  addMessage: PropTypes.func.isRequired,
};

const mapStateToProps = ({watsonResponse}) => ({ response: watsonResponse.response });

const mapDispatchToProps = dispatch => {
  return {
    postMessage: bindActionCreators(postMessage, dispatch)
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Chat);
