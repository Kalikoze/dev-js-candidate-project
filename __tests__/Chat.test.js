import React from 'react';
import Chat from '../components/Chat';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { mount, configure } from 'enzyme';

configure({ adapter: new Adapter() });

describe('Chat component', () => {
  let wrapper;
  const mockFn = jest.fn();

  beforeEach(() => {
    wrapper = mount(<Chat addMessage={mockFn} />);
  });

  it('should render the Chat component correctly with the right classes and styles', () => {
    const tree = renderer.create(wrapper).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be defined', () => {
    expect(wrapper).toBeDefined();
  });

  it('should update state as message is typed out in input', () => {
    const chatInput = wrapper.find('input');

    expect(wrapper.state().input).toEqual('');
    chatInput.simulate('change', { target: { value: 'Hello SpruceBot!'} });
    expect(wrapper.state().input).toEqual('Hello SpruceBot!');
  });

  it.skip('should rest input of state to an empty string after clicking submit', () => {
    const chatInput = wrapper.find('input');
    const submit = wrapper.find('button');

    chatInput.simulate('change', { target: { value: 'Hello SpruceBot!'} });
    expect(wrapper.state().input).toEqual('Hello SpruceBot!');

    // Unfortunately neither of these are working the way I expect them too
    chatInput.simulate('keypress', { key: 'Enter', keyCode: 13, which: 13 });
    submit.simulate('click');
    expect(wrapper.state().input).toEqual('');
  });

  it('should have addMessage function as a prop', () => {
    expect(wrapper.props().addMessage).toEqual(mockFn);
  });
});
