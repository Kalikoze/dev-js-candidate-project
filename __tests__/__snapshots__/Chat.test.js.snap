// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component should render the Chat component correctly with the right classes and styles 1`] = `
<section
  className="jsx-799177478 input-section"
>
  <input
    className="jsx-799177478"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Chat with SpruceBot here..."
    value=""
  />
  <button
    className="jsx-799177478"
    onClick={[Function]}
  />
</section>
`;
